Consistent locale usage for number representation

m-kuhn opened the pull request:
Followup on discussion in https://github.com/qgis/QGIS/pull/6833

@elpaso this reverts some code where the user locale is used for number representation while everything else uses the system locale for this. What was it that looked different and led you to 
https://github.com/qgis/QGIS/commit/8ae65b3c70291a19c440a76172ac02bdfcb27c86#diff-327a6927d0fab68c4c6cb99928081e9cR35

m-kuhn committed:
Use system locale for number representation

m-kuhn commented:
According to the commit message it's cosmetic. https://github.com/qgis/QGIS/commit/8ae65b3c70291a19c440a76172ac02bdfcb27c86#diff-327a6927d0fab68c4c6cb99928081e9cR35

>  [bugfix] Range formatter for doubles and ints
> 
> Double and ints are now formatted according
to chosen locale (in user settings) and precision
(from the widget field config)
>
> Numbers are also right aligned in the form
and not only in the table.
>
> IMHO a bugfix since the numbers looked very
different in the form and the table
when the system locale used decimal
separator and group separator which were not
the same than system locale.

I guess this should be solved on a different level by either killing group separators application wide or fixing group separators in wherever they didn't correspond to the system locale. Let's wait for @elpaso

elpaso commented:
@m-kuhn +1 to solve this at different level, what exactly do you have in mind?

IMHO we should try to respect decimal separator and currency from the user-defined locale, or at least provide an option for setting them independently.

In all cases the numbers should look the same in the form and in the attribute table (and the identify)

m-kuhn commented:
The way to do this consistently, would bo to overwrite the system locale.

What I am mainly wondering is where 8ae65b3 solved incosistencies. Before 8ae65b3 I don't see where the user locale has been used for number representation. So this pull request here proposes to revert to only use system locale for number representation but at least consistently.

elpaso commented:
Did you have a look to the PR https://github.com/qgis/QGIS/pull/6246 ? The video shows what id does: , right now I can't remember exactly what the inconsistency was, I think it was in the spinbox /range widget formatting.

m-kuhn commented:
Just now I had a look at it, thanks for the link.

However, with my test data here it's both consistent, with and without 8ae65b3. The difference is that with 8ae65b3 it looks like this on my en_US system with qgis set to Italian:

![image](https://user-images.githubusercontent.com/588407/39533686-017dd43a-4e30-11e8-95bb-a0675c25fbee.png)

![image](https://user-images.githubusercontent.com/588407/39533692-06b214a2-4e30-11e8-8ca4-ad805f2aa196.png)

![image](https://user-images.githubusercontent.com/588407/39533705-0df6a836-4e30-11e8-9c9d-f981762448cb.png)

![image](https://user-images.githubusercontent.com/588407/39533718-174636b8-4e30-11e8-876a-cec0dd1afc6a.png)

elpaso commented:
@m-kuhn we need to decide what we is our goal here: if we want to display values according to the locale that the user selected in QGIS (or provide more fine-grained control over currency and decimal separator) there is still a lot of work to do but my original PR was a first step in the right direction, I know that is wasn't a complete fix.
If you want to revert and propose a complete solution based on a different approach I'm totally fine with that.

m-kuhn commented:
I think in the first place this pull request fixes a regression.

Then I am not opposed to introducing more control in a second step.

jef-n commented:
Why do we need to override the system locale at all (except for the cases where we need locale C for external stuff)?  The user should already have control over the settings outside of QGIS.  Selecting the translation is one thing - but only because you might prefer no translation over one of our incomplete ones.

elpaso commented:
@jef-n you are right, but I know a lot of people (and I'm one of those) that keeps the italian locale at the operating system level (for time and money) but hates `,` as a decimal separator, for this reason I usually choose english instead.

If we could have a more fine-grained control, that would be awesome.

andreasneumann commented:
@elpaso - I am in the same boat. I hate the comma (,) as decimal separator. Other than that, the german locale is fine for me. And that is basically the main difference between the german and swiss locale (except maybe the € as default currency and the scharfes "s" vs double "s".

nyalldawson commented:
Clueless outsider here:

Does any other program offer a setting for this? E.g. Excel? Libreoffice Calc? Or do they all just say "you get what the locale gives you"?

m-kuhn commented:
> @nyalldawson all spreadsheets and word processors do.

They do offer or they do say "you get what the locale gives you"?

andreasneumann commented:
@m-kuhn 
Excel offers it as a setting - separate from the locale. Same in Libreoffice. With the option to inherit it from the operating system or overwrite it. Both, for the thousand separator or the decimal point separator.

![image](https://user-images.githubusercontent.com/884476/39564751-c78d7d3a-4eb5-11e8-80df-8e08d3bd96da.png)

andreasneumann commented:
actually libreoffice has a bit fewer options compared to Excel. But it does allow to override the locale setting.

andreasneumann commented:
libre office setting:

![image](https://user-images.githubusercontent.com/884476/39564895-4416d05e-4eb6-11e8-89d2-14ce4373af83.png)

it has separate settings for regional settings, separated from the language of the GUI, and allows to use the regional settings for decimal separator or not.

No separate option for thousand separator (except through regional setting).

elpaso commented:
After a quick chat with @m-kuhn we believe that the best approach here is:

- merge this PR
- immediately rename "Override system locale" to "Override system language" or "Override system translation" (whatever is clearer in english) 

for the future:

- have another option to override default (system locale) for (at least):
- LC_NUMERIC
- LC_TIME

m-kuhn committed:
Explicitly refer to translation and not locale on override

We only change the user interface translation and do not touch the locale for other things like number representation.

m-kuhn committed:
Use default (not system) locale

m-kuhn committed:
Adjust tests

m-kuhn commented:
Thanks a lot everyone. And please let me know if there are other inconsistencies (re)introduced.
